### PR TITLE
Ensure Collaboration Workspace is able to render when matching redirects

### LINKED ISSUE #93: Integrate collaboration service with matching service
Collaboration service still uses dummy matching api wrappers, so we have to integrate it and make everything work together now :)

### PR DESCRIPTION
**Pull request by tryyang2001**
# Pull Request

## Description

I have updated the collab context function to ensure that the collaboration workspace page is able to render. I also did a little bit of code refactoring, and removed unnecessary console.log.

I also fixed the unique key issue that ly faced, and thus removed the message uuid and the uuid package. 

There are still some works left undone, please refer to the additional notes.

## Related Issue(s)

Fixes #93, #95 

## Changes Made

- update `handleConnectToRoom` in `collab.tsx`
- create a `hashUtils` for verifying room id checksum
- added a `isNotFound` state value in the collab context, the reason is that if we simply throw `notFound()` in `handleConnectToRoom`, the error will not be reflected in the collaboration room page, hence we need this error state to reflect the 404
- code clean up, removed some console log, first to improve efficiency (we don't want to show the code changes during collaboration as it will cause latency, and we also don't want to show the message sent in the console every time we sent a message), next is for cleanliness (we don't want to print a random user id, or whatever id in the console)
- fixed the height issue in the dashboard page

## Screenshots (if applicable)

Not sure how to show the result, but now the below page uses the values returned from matching, and it now supports concurrent update (thanks @weidak )
image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References

I want to list out a few things that we need to do:
1. We might need a way to show the partner connection status, so that we know if our partner loses the connection.
2. We need a way to enforce the 2-hour limit for our collaboration workspace. This might need to use database to store the remaining time, so that even if the user closes the browser and joined back, the time is still correct.
3. We need to indicate that there is a new message coming when one sent an message. We likely want to use the [Badge ](https://nextui.org/docs/components/badge)in Next-Ui.
4. When the user left the session by closing the browser and joined back the room, the chat history is lost, we might need to restore the chat history in the future. This will be one of the requirements in the history service.
5. Same for code. When the user left and joined back the room, he will not be able to see the latest code, but the default code only.

More issues are coming...
6. We need a way to standardise the size of each user avatar, this is for better UI:
As we can see here, the "Mr Test" profile is more in ellipse shape while our "John Doe" profile is more in circular shape. 
image
image
This is quite obvious when we run a match, and it can go even worse if one image is much higher/wider than the other.
image

7. For matching, right now we might not have all the questions available for each topic that the user chose, I think we need to add more questions.

### CODE REVIEW

File: frontend/src/helpers/collaboration/socket_service.ts
Comment (weidak) [inline frontend/src/helpers/collaboration/socket_service.ts:1]: Thanks for removing the console log 🙏🏻 😓

File: frontend/src/components/collab/chat/ChatSpace.tsx [hunk 1]
```diff
@@ -92,7 +88,11 @@ const ChatSpace = ({ onClose }: IChatSpaceProps) => {
         {
           <ul className="space-y-3 px-4">
             {messages.map((item) => (
-              <ChatBubble message={item} isSelf={item.senderId === user.id!} />
+              <ChatBubble
+                key={crypto.randomUUID()}
```
Comment (weidak) [inline frontend/src/components/collab/chat/ChatSpace.tsx:92]: oh u got it to work

File: frontend/src/utils/hashUtils.ts
Comment (weidak) [inline frontend/src/utils/hashUtils.ts:1]: Looks good!

**Review by weidak — APPROVED** [review]

**Comment by weidak** [comment]
> We might need a way to show the partner connection status, so that we know if our partner loses the connection.

Ok! should be quite doable. 

> We need to indicate that there is a new message coming when one sent an message. We likely want to use the [Badge ](https://nextui.org/docs/components/badge)in Next-Ui. 

Ok, this one doesnt sound too hard

> We need a way to enforce the 2-hour limit for our collaboration workspace. This might need to use database to store the remaining time, so that even if the user closes the browser and joined back, the time is still correct.
> When the user left the session by closing the browser and joined back the room, the chat history is lost, we might need to restore the chat history in the future. This will be one of the requirements in the history service.
Same for code. When the user left and joined back the room, he will not be able to see the latest code, but the default code only.

This two points should be ok, i was playing around with redis this afternoon and i believe we can have a middleware that retrieves the data from the session and set up the room the way it was before the user left. During the exchange of any events, I will implement it such that it keeps writing to the Redis cache. I think the history service part comes when the session is complete/ended, and the cache then pushes the finished code to the history service table together with the questionid such that a user can retrieve his previous attempts. 

I will try to finish it by Sunday, today my com crashed and i was busy reformatting/reinstalling things (i lost my progress too shd have just pushed to github hais)

Anyways, this looks good Rui yang! I will take over from here and u can focus on ur other mods 👍🏻

**Comment by tlyi** [comment]
> I also fixed the unique key issue that ly faced, and thus removed the message uuid and the uuid package.

Thank you~~

> 3. We need to indicate that there is a new message coming when one sent an message. We likely want to use the [Badge ](https://nextui.org/docs/components/badge)in Next-Ui.

Oh yes was thinking of doing this! I'll do it in the next sprint :) 

> 4. When the user left the session by closing the browser and joined back the room, the chat history is lost, we might need to restore the chat history in the future. This will be one of the requirements in the history service.

Personally I feel like it is not important to restore chat history after the session/as a user I would rather not have my chat history stored for future reference xD

However, I agree that we should implement caching for the chat history while the session is ongoing :)

Overall, looks neat! Thanks for tidying things up too ry! 👍

**Comment by tryyang2001** [comment]
> Personally I feel like it is not important to restore chat history after the session

I agree with you that we don't really need to restore the chat history after the session. But how about during the session? For example A and B are connected, and then halfway through A left and then joined back to the room, shall we restore the chat history for A?

File: frontend/src/components/collab/chat/ChatSpace.tsx (hunk 1 above)
Comment (tryyang2001) [reply frontend/src/components/collab/chat/ChatSpace.tsx:92]: I think the issue is that when we provide elements inside a `ul` tag, we need to explicitly provide a key.

**Comment by tlyi** [comment]
> I agree with you that we don't really need to restore the chat history after the session. But how about during the session? For example A and B are connected, and then halfway through A left and then joined back to the room, shall we restore the chat history for A?

Yea, that's what I was referring to when I said "we should implement caching for the chat history while the session is ongoing" :D However, the way I am currently sending the chats now is message by message so that the component adds to the current messages instead of re-rendering the whole chat, so there is no way for the other user (A) to retrieve the message that the other party (B) has sent while A is disconnected from the socket.

These are some different approaches I can think of:
1. Send the whole chat history through the socket so that when a new message is sent, the entire new array of messages is sent through the socket. This way, the disconnected user A can retrieve the chat history when he rejoins.
- Issue: Not sure how it would affect the chat rendering + when the chat gets too long, I am not sure if it is a good idea to send the whole thing through the socket
2. Chats are cached in a centralised cache
- Issue: I'm not sure how to use Redis, but if @weidak manages to do it for the code editor, then I can refer :D 

I'm leaning towards Approach 2 because this way, it is also possible for me to keep track of what messages each party has "read" (not just received), which makes it easier to implement the notification thing :)

**Comment by tryyang2001** [comment]
> Yea, that's what I was referring to when I said "we should implement caching for the chat history while the session is ongoing"

I see! For the implementation details, I highly recommend that we use Redis database to store the chat history when the session is still active, and delete the chat storage when the session ended. I also think that approach 2 that you mentioned is more elegant. Additionally, the approach 2 and 3 are actually the same. "Cache" is just another word we used to describe a database, and specifically a fast read/write database. An example will be Redis, which is nothing special but a normal database that we can use, and is pretty fast compared to the other database. So I think we can create a Redis server soon to implement this restoring chat history feature.

**Review by weidak — APPROVED** [review]

**Pull request merged**